utilities/fault_injection_fs.cc

```diff
@@ -1093,51 +1201,130 @@ bool FaultInjectionTestFS::TryParseFileName(const std::string& file_name,
-    allowed_type = true;
-  } else {
-    uint64_t number;
-    FileType cur_type = kTempFile;
-    if (TryParseFileName(file_name, &number, &cur_type)) {
-      for (const auto& type : write_error_allowed_types_) {
-        if (cur_type == type) {
-          allowed_type = true;
-        }
-      }
+  IOStatus ret;
+  if (ctx->rand.OneIn(ctx->one_in)) {
+    ctx->count++;
+    if (ctx->callstack) {
+      free(ctx->callstack);
     }
+    ctx->callstack = port::SaveStack(&ctx->frames);
+    ctx->message = "injected write error";
+    ret = IOStatus::IOError(ctx->message);
+    ret.SetRetryable(ctx->retryable);
+    ret.SetDataLoss(ctx->has_data_loss);
+  }
+  return ret;
+}
+
+IOStatus FaultInjectionTestFS::MaybeInjectThreadLocalMetadataReadError(
+    const IOOptions& io_options) {
+  ErrorContext* ctx = static_cast<ErrorContext*>(
+      injected_thread_local_metadata_read_error_->Get());
+  if (ctx == nullptr || !ctx->enable_error_injection || !ctx->one_in ||
+      io_activties_exempted_from_fault_injection.find(io_options.io_activity) !=
+          io_activties_exempted_from_fault_injection.end()) {
+    return IOStatus::OK();
   }
 
-  if (allowed_type) {
-    if (write_error_rand_.OneIn(write_error_one_in_)) {
-      return GetError();
+  IOStatus ret;
+  if (ctx->rand.OneIn(ctx->one_in)) {
+    ctx->count++;
+    if (ctx->callstack) {
+      free(ctx->callstack);
     }
+    ctx->callstack = port::SaveStack(&ctx->frames);
+    ctx->message = "injected metadata read error";
+    ret = IOStatus::IOError(ctx->message);
+    ret.SetRetryable(ctx->retryable);
+    ret.SetDataLoss(ctx->has_data_loss);
   }
-  return IOStatus::OK();
+  return ret;
 }
 
-IOStatus FaultInjectionTestFS::InjectMetadataWriteError() {
-  {
-    MutexLock l(&mutex_);
-    if (!enable_metadata_write_error_injection_ ||
-        !metadata_write_error_one_in_ ||
-        !write_error_rand_.OneIn(metadata_write_error_one_in_)) {
-      return IOStatus::OK();
+IOStatus FaultInjectionTestFS::MaybeInjectThreadLocalMetadataWriteError(
+    const IOOptions& io_options) {
+  ErrorContext* ctx = static_cast<ErrorContext*>(
+      injected_thread_local_metadata_write_error_->Get());
+  if (ctx == nullptr || !ctx->enable_error_injection || !ctx->one_in ||
+      io_activties_exempted_from_fault_injection.find(io_options.io_activity) !=
+          io_activties_exempted_from_fault_injection.end()) {
+    return IOStatus::OK();
+  }
+
+  IOStatus ret;
+  if (ctx->rand.OneIn(ctx->one_in)) {
+    ctx->count++;
+    if (ctx->callstack) {
+      free(ctx->callstack);
     }
+    ctx->callstack = port::SaveStack(&ctx->frames);
+    ctx->message = "injected metadata write error";
+    ret = IOStatus::IOError(ctx->message);
+    ret.SetRetryable(ctx->retryable);
+    ret.SetDataLoss(ctx->has_data_loss);
+    TEST_SYNC_POINT("FaultInjectionTestFS::InjectMetadataWriteError:Injected");
+  }
+  return ret;
+}
+
+void FaultInjectionTestFS::PrintInjectedThreadLocalReadErrorBacktrace() {
```

Comment:
Copy-paste that can easily use a helper taking a reference.

Reply:
Fixed